core/embed/rust/src/ui/api/firmware_upy.rs

```diff
@@ -0,0 +1,1065 @@
+use crate::{
+    io::BinaryData,
+    micropython::{
+        gc::Gc,
+        iter::IterBuf,
+        list::List,
+        macros::{obj_fn_1, obj_fn_kw, obj_module},
+        map::Map,
+        module::Module,
+        obj::Obj,
+        qstr::Qstr,
+        util,
+    },
+    strutil::TString,
+    trezorhal::model,
+    ui::{
+        backlight::BACKLIGHT_LEVELS_OBJ,
+        component::Empty,
+        layout::{
+            base::LAYOUT_STATE,
+            obj::{ComponentMsgObj, LayoutObj, ATTACH_TYPE_OBJ},
+            result::{CANCELLED, CONFIRMED, INFO},
+            util::{upy_disable_animation, RecoveryType},
+        },
+        ui_features::ModelUI,
+        ui_features_fw::UIFeaturesFirmware,
+    },
+};
+use heapless::Vec;
+
+/// Dummy implementation so that we can use `Empty` in a return type of unimplemented trait
+/// function
+impl ComponentMsgObj for Empty {
+    fn msg_try_into_obj(&self, _msg: Self::Msg) -> Result<Obj, crate::error::Error> {
+        Ok(Obj::const_none())
+    }
+}
+
+// free-standing functions exported to MicroPython mirror `trait
+// UIFeaturesFirmware`
+// NOTE: `disable_animation` not a part of trait UiFeaturesFirmware
+
+extern "C" fn new_confirm_action(n_args: usize, args: *const Obj, kwargs: *mut Map) -> Obj {
```

Comment:
Would this be a good time to drop the `new_` prefix? Or is there a reason to still keep it around?

Reply:
I'd skip this for now because it would make the rebase complicated. 
But maybe later in a separate commit.